### PR TITLE
Sanitize prometheus metric labels

### LINKED ISSUE #7367: bosh-deployment is not a valid label name, prometheus metrics
## Summary

Upgrading to Concourse v7.4.0, I receive the following errors on our web nodes:

```
[...]
panic: descriptor Desc{fqName: "concourse_error_logs", help: "Number of error logged", constLabels: {}, variableLabels: [message]} is invalid: "bosh-deployment" is not a valid label name for metric "concourse_error_logs"

goroutine 1 [running]:
github.com/prometheus/client_golang/prometheus.(*Registry).MustRegister(0xc0000a4ff0, 0xc0007b28b0, 0x1, 0x1)
        github.com/prometheus/client_golang@v1.11.0/prometheus/registry.go:403 +0xb7
github.com/prometheus/client_golang/prometheus.MustRegister(...)
        github.com/prometheus/client_golang@v1.11.0/prometheus/registry.go:178
github.com/concourse/concourse/atc/metric/emitter.(*PrometheusConfig).NewEmitter(0xc000518ce0, 0xc000779b60, 0x0, 0x0, 0x1, 0xc0007b2850)
        github.com/concourse/concourse/atc/metric/emitter/prometheus.go:133 +0x184
github.com/concourse/concourse/atc/metric.(*Monitor).Initialize(0xc00009e000, 0x36cca90, 0xc00053dc80, 0xc0007a26f0, 0x24, 0xc000779b60, 0x3e8, 0xc00053dc80, 0xc000320340)
        github.com/concourse/concourse/atc/metric/emit.go:143 +0x4c5
github.com/concourse/concourse/atc/atccmd.(*RunCommand).configureMetrics(0xc0001d8b00, 0x36cca90, 0xc00053db00, 0x36614c0, 0x335b5d8)
        github.com/concourse/concourse/atc/atccmd/command.go:1599 +0xd5
github.com/concourse/concourse/atc/atccmd.(*RunCommand).Runner(0xc0001d8b00, 0xc00079be20, 0x0, 0x1, 0x0, 0x0, 0x0, 0x0)
        github.com/concourse/concourse/atc/atccmd/command.go:571 +0x545
main.(*WebCommand).Runner(0xc0002d4008, 0xc00079be20, 0x0, 0x1, 0x2, 0x6, 0xc00079bc90, 0xc00028fd00)
        github.com/concourse/concourse/cmd/concourse/web.go:62 +0xa5
main.(*WebCommand).Execute(0xc0002d4008, 0xc00079be20, 0x0, 0x1, 0x2778480, 0x2af9320)
        github.com/concourse/concourse/cmd/concourse/web.go:44 +0x65
github.com/vito/twentythousandtonnesofcrudeoil.installEnv.func2(0x7fa6563cdcc0, 0xc0002d4008, 0xc00079be20, 0x0, 0x1, 0x1, 0x182)
        github.com/vito/twentythousandtonnesofcrudeoil@v0.0.0-20180305154709-3b21ad808fcb/environment.go:40 +0x8a
github.com/jessevdk/go-flags.(*Parser).ParseArgs(0xc00027d880, 0xc000130010, 0x1, 0x1, 0xc00015e640, 0x0, 0x0, 0x0, 0x0)
        github.com/jessevdk/go-flags@v1.4.1-0.20200711081900-c17162fe8fd7/parser.go:340 +0x85d
github.com/jessevdk/go-flags.(*Parser).Parse(...)
        github.com/jessevdk/go-flags@v1.4.1-0.20200711081900-c17162fe8fd7/parser.go:190
main.main()
        github.com/concourse/concourse/cmd/concourse/main.go:30 +0x1f5
```

Looks like it's trying to specify an invalid label, I believe dashes aren't allowed in label names for Prometheus.

## Steps to reproduce

Using this bosh config when deploying:

```yaml
name: concourse
instance_groups:
[...]
- azs:
  - AZ1
  instances: 4
  jobs:
  - name: web
    properties:
      add_local_users: [...]
      audit:
        build: true
        job: true
        pipeline: true
        system: true
        team: true
      auth_duration: 48h
      bind_ip: 127.0.0.1
      bind_port: 8080
      build_log_retention:
        default_builds: 20
        default_days: 30
        maximum_builds: 50
        maximum_days: 60
      build_tracker_interval: 10s
      capture_error_metrics: true
      container_placement_strategies:
      - limit-active-tasks
      - volume-locality
      - fewest-build-containers
      default_check_interval: 2m
      default_resource_type_check_interval: 30m
      default_task_cpu_limit: 150
      default_task_memory_limit: 4096mb
      enable_archive_pipeline: true
      enable_cache_streamed_volumes: true
      enable_global_resources: false
      enable_pipeline_instances: true
      enable_rerun_when_worker_disappears: true
      enable_skip_checking_not_in_use_resources: true
      external_url: https://[...]
      generic_oidc:
        client_id: [...]
        client_secret: [...]
        display_name: Single SignOn
        groups_key: groups
        issuer: [...]
        scopes: https://graph.microsoft.com/.default profile email openid
        user_name_key: email
      global_resource_check_timeout: 10m
      job_scheduling_max_in_flight: 8
      ldap_auth:
        bind_dn: [...]
        bind_pw: [...]
        display_name: [...]
        group_search_base_dn: [...]
        group_search_filter: [...]
        group_search_group_attr: member
        group_search_name_attr: cn
        group_search_scope: sub
        group_search_user_attr: distinguishedName
        host: [...]
        insecure_skip_verify: true
        start_tls: false
        user_search_base_dn: [...]
        user_search_email_attr: mail
        user_search_filter: (objectClass=person)
        user_search_id_attr: sAMAccountName
        user_search_name_attr: displayName
        user_search_scope: sub
        user_search_username: sAMAccountName
      lidar_scanner_interval: 2m
      main_team:
        auth:
          ldap:
            groups: [...]
            users: []
          local:
            users: [...]
          oidc:
            groups: [...]
      max_active_tasks_per_worker: 10
      p2p:
        enable: true
        stream_timeout: 15m
      postgresql:
        database: atc
        role:
          name: [...]
          password: [...]
      prometheus:
        bind_ip: 0.0.0.0
        bind_port: 8456
      redact_secrets: false
      secrets:
        cache:
          duration: 2m
          duration_notfound: 2m
          enabled: true
          purge_interval: 10m
        retry_attempts: 3
        retry_interval: 30s
      streaming_artifacts_compression: gzip
      token_signing_key:
        private_key: [...]
        public_key: [...]
      vault:
        auth:
          client_token: [...]
        path_prefix: /concourse
        query_timeout: 30s
        shared_path: /concourse-template
        url: [...]
      worker_gateway:
        authorized_keys: [...]
        host_key:
          private_key: [...]
          public_key: [...]
    release: concourse
  name: web
  networks:
  - default:
    - dns
    - gateway
    name: default
    static_ips: [...]
  stemcell: default
  vm_type: cc-app-web
releases:
- name: "concourse"
  version: "7.4.0"
  url: "https://bosh.io/d/github.com/concourse/concourse-bosh-release?v=7.4.0"
  sha1: "a0737a17fc46490ede1e83516f02a92c009417a6"
- name: bpm
  sha1: 502e9446fa34accaf122ad2b28b6ffa543d5bbca
  url: https://bosh.io/d/github.com/cloudfoundry/bpm-release?v=1.1.12
  version: 1.1.12
- name: postgres
  sha1: e44bbe8f8a7cdde1cda67b202e399a239d104db6
  url: https://bosh.io/d/github.com/cloudfoundry/postgres-release?v=43
  version: "43"
- name: syslog
  sha1: 3c6636831f3c89fa9e8f4b34613b5cefd3b29046
  url: https://bosh.io/d/github.com/cloudfoundry/syslog-release?v=11.7.5
  version: 11.7.5
- name: nginx
  sha1: 7850fda50d355316c49d16a4e1c4d30227204733
  url: https://bosh.io/d/github.com/cloudfoundry-community/nginx-release?v=1.19.10
  version: 1.19.10
- name: os-conf
  sha1: 4f653168954749992a541d228dd4f936f2eff2d6
  url: https://bosh.io/d/github.com/cloudfoundry/os-conf-release?v=22.1.1
  version: 22.1.1
- name: cron
  sha1: f0a048b04fa8f064a6170b34741439f91e7838b5
  url: https://bosh.io/d/github.com/cloudfoundry-community/cron-boshrelease?v=1.3.1
  version: 1.3.1
- name: node-exporter
  sha1: cea1edc938873b7204b889173002e0e9fa813e54
  url: https://bosh.io/d/github.com/cloudfoundry-community/node-exporter-boshrelease?v=5.0.0
  version: 5.0.0
- name: bind9
  sha1: 8a6a99046ea1adba07e0a95d600e7f608f5bb305
  url: https://bosh.io/d/github.com/cloudfoundry-community/bind9-boshrelease?v=1.3.0
  version: 1.3.0
```

## Expected results

Expect it to deploy successfully, while also exporting metrics with the bosh deployment being identified.

## Actual results

Due to the error_metrics being enabled, and it trying to specify the "bosh-deployment" label, which isn't a valid Prometheus label, it fails.

## Triaging info

* Concourse version: v7.4.0 via Bosh
* Did this used to work? Yes, however I don't think bosh-deployments was an exported label in previous versions.

### PR DESCRIPTION
**Pull request by lrstanley**
## Changes proposed by this PR

closes #7367

* [x] sanitize prometheus labels to ensure any characters not supported by prometheus are replaced by `_`'s. Underscores are also  stripped from the start/end, as well as if there are any duplicate underscore characters.

## Notes to reviewer

Used the following environment variables in my docker-compose.yml to replicate the issues seen in #7367:

```console
      CONCOURSE_METRICS_ATTRIBUTE: "bosh-deployment:concourse-deployment,bosh-job:concourse-job"
      CONCOURSE_CAPTURE_ERROR_METRICS: "true"
      CONCOURSE_PROMETHEUS_BIND_IP: "0.0.0.0"
      CONCOURSE_PROMETHEUS_BIND_PORT: "8456"
```

Didn't fix this on the bosh side, as those metric names have existed for other emitters for awhile now, and changing the name may break metrics for users.

## Release Note

* Ensure Prometheus metric labels are valid. This resolves an issue with our bosh release, where web nodes would fail to start, due to a metric label that wasn't valid according to Prometheus.

### CODE REVIEW

**Comment by xtremerui** [comment]
Thank you for the PR. If the changes in go.mod and package.json are not necessary could you please undo that to resolve the conflict?

**Comment by lrstanley** [comment]
> Thank you for the PR. If the changes in go.mod and package.json are not necessary could you please undo that to resolve the conflict?

Yeah, those weren't supposed to be added, will correct it shortly. Unfortunately, I enabled dependency auto-correction on my github account for all new repositories by default, and that seems to include clones... which is a bit annoying.

**Comment by lrstanley** [comment]
Should be fixed. 👍🏻

File: atc/metric/emitter/prometheus_test.go [hunk 1]
```diff
@@ -188,7 +188,9 @@ var _ = Describe("PrometheusEmitter", func() {
 	})
 
 	JustBeforeEach(func() {
-		prometheusEmitter, err = prometheusConfig.NewEmitter(nil)
+		prometheusEmitter, err = prometheusConfig.NewEmitter(map[string]string{
+			"invalid-label": "foo", // Ensure invalid labels are sanitized.
```
Comment (xtremerui) [inline atc/metric/emitter/prometheus_test.go:192]: Could you please add some more test strings here e.g. _blah__

Just to cover the regex `rePrometheusLabelClean`.

**Comment by xtremerui** [comment]
The test seems failed here https://ci.concourse-ci.org/teams/contributor/pipelines/pr/jobs/unit/builds/7.1?vars.number=7423#L611bacf6:329

**Comment by lrstanley** [comment]
Tests should be fixed, not sure how I missed it, sorry!

**Review by xtremerui — APPROVED** [review]
LGTM thx for your patience!

**Pull request merged**